lib/web/join_tokens.go

```diff
@@ -572,134 +522,46 @@ func getJoinScript(ctx context.Context, settings scriptSettings, m nodeAPIGetter
-	// This ensures the initial installed version is the same as the `teleport-ent-updater` would install.
-	if settings.installUpdater {
-		if settings.automaticUpgradesVersion == "" {
-			return "", trace.Wrap(err, "automatic upgrades version must be set when installUpdater is true")
-		}
-
-		repoChannel = stableCloudChannelRepo
-		// automaticUpgradesVersion has vX.Y.Z format, however the script
-		// expects the version to not include the `v` so we strip it
-		version = strings.TrimPrefix(settings.automaticUpgradesVersion, "v")
-	}
-
-	// This section relies on Go's default zero values to make sure that the settings
-	// are correct when not installing an app.
-	err = scripts.InstallNodeBashScript.Execute(&buf, map[string]interface{}{
-		"token":    settings.token,
-		"hostname": hostname,
-		"port":     portStr,
-		// The install.sh script has some manually generated configs and some
-		// generated by the `teleport <service> config` commands. The old bash
-		// version used space delimited values whereas the teleport command uses
-		// a comma delimeter. The Old version can be removed when the install.sh
-		// file has been completely converted over.
-		"caPinsOld":                  strings.Join(caPins, " "),
-		"caPins":                     strings.Join(caPins, ","),
-		"packageName":                packageName,
-		"repoChannel":                repoChannel,
-		"installUpdater":             strconv.FormatBool(settings.installUpdater),
-		"version":                    shsprintf.EscapeDefaultContext(version),
-		"appInstallMode":             strconv.FormatBool(settings.appInstallMode),
-		"appServerResourceLabels":    appServerResourceLabels,
-		"appName":                    shsprintf.EscapeDefaultContext(settings.appName),
-		"appURI":                     shsprintf.EscapeDefaultContext(settings.appURI),
-		"joinMethod":                 shsprintf.EscapeDefaultContext(settings.joinMethod),
-		"labels":                     strings.Join(labelsList, ","),
-		"databaseInstallMode":        strconv.FormatBool(settings.databaseInstallMode),
-		"db_service_resource_labels": dbServiceResourceLabels,
-		"discoveryInstallMode":       settings.discoveryInstallMode,
-		"discoveryGroup":             shsprintf.EscapeDefaultContext(settings.discoveryGroup),
-	})
+	installOpts, err := h.installScriptOptions(ctx)
 	if err != nil {
-		return "", trace.Wrap(err)
-	}
-
-	return buf.String(), nil
+		return "", trace.Wrap(err, "Building install script options")
+	}
+
+	nodeInstallOpts := scripts.InstallNodeScriptOptions{
+		InstallOptions: installOpts,
+		Token:          token.GetName(),
+		CAPins:         caPins,
+		// We are using the joinMethod from the script settings instead of the one from the token
+		// to reproduce the previous script behavior. I'm also afraid that using the
+		// join method from the token would provide an oracle for an attacker wanting to discover
+		// the join method.
```

Comment:
Note: an attacker knowing a token can already discover its labels today, so I don't think it's that bad if we also set the join method.